### PR TITLE
IA-2066: Adding Trace Id to Leonardo routes header

### PR DESCRIPTION
**Pull request by xuscrbw**
Adding some tracing to our logs. With a trace id, it should be easier to determine which routes have produced which log entries. It looks like Rob is also doing updates in https://github.com/DataBiosphere/leonardo/pull/1685
---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeStatusTransitionsSpec.scala [hunk 1]
```diff
@@ -13,8 +13,12 @@ import org.scalatest.{DoNotDiscover, ParallelTestExecution}
  * Note these tests can take a long time so we don't test all edge cases, but these cases
  * should exercise the most commonly used paths through the system.
  */
-@DoNotDiscover
-class RuntimeStatusTransitionsSpec extends GPAllocFixtureSpec with ParallelTestExecution with LeonardoTestUtils {
+//@DoNotDiscover
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimeStatusTransitionsSpec.scala]: Reminder to revert this before merge

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/models.scala [hunk 2]
```diff
@@ -28,4 +28,13 @@ object AppContext {
     } yield ApplicativeAsk.const[F, AppContext](
       context
     )
+
+  def withTraceId[F[_]: Sync: Timer](
```
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/models.scala]: Is this not equivalent to `lift` (line 25)?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/AppRoutes.scala [hunk 3]
```diff
@@ -19,154 +19,156 @@ import org.broadinstitute.dsde.workbench.leonardo.http.api.AppRoutes._
...
-              complete(
-                listAppHandler(userInfo, None, params)
-              )
-            }
-          }
-        } ~
-          pathPrefix(googleProjectSegment) { googleProject =>
+class AppRoutes(kubernetesService: KubernetesService[IO], userInfoDirectives: UserInfoDirectives) {
+
+  val routes: server.Route = traceRequestForService(serviceData) { span =>
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/AppRoutes.scala:25]: I was wondering whether `traceRequestForService` could be at the top-level, or if it should be done per-route. I guess it should be okay here.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala [hunk 4]
```diff
@@ -28,250 +26,293 @@ import org.broadinstitute.dsde.workbench.leonardo.http.service.{
...
-  }
-
-  private[api] def getRuntimeHandler(userInfo: UserInfo,
-                                     googleProject: GoogleProject,
-                                     runtimeName: RuntimeName,
-                                     span: Span): IO[ToResponseMarshallable] = {
-    val res = for {
-      implicit0(ctx: ApplicativeAsk[IO, AppContext]) <- AppContext.lift[IO](Some(span))
-      resp <- runtimeService.getRuntime(userInfo, googleProject, runtimeName)
+//    val res = for {
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala]: Reminder to remove commented code

**Comment by rtitle** [comment]
Couple things to clean up, but looks great! Looks like I duplicated some things in https://github.com/DataBiosphere/leonardo/pull/1685, haha. I'm happy to rebase my PR -- I can use it to add counters, and spans/counters to `ProxyRoutes` (which this PR doesn't touch).

cc @Qi77Qi

**Review by rtitle — APPROVED** [review]
This looks good to me! Nice work

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/package.scala [hunk 5]
```diff
@@ -45,6 +52,17 @@ package object api {
         case None    => d2
       }
   }
+
+  def extractAppContext(span: Option[Span]): Directive1[ApplicativeAsk[IO, AppContext]] =
+    //val traceIdFromHeader = req.headers.get(CaseInsensitiveString("X-Cloud-Trace-Context")).map(x => TraceId(x.value))
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/package.scala:57]: can you remove the commented out code

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/package.scala:57]: Actually just did here: https://github.com/DataBiosphere/leonardo/pull/1685
